### PR TITLE
#655 FIX

### PR DESCRIPTION
**Pull request by TikhomirovSergey**
## Change list

#655 FIX
 
## Types of changes

- [ ] No changes in production code.
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Details

`io.appium.java_client.remote.AppiumCommandExecutor` was rolled back.

It is breaking change because it is compatible with appium server > v1.6.5

### CODE REVIEW

**Comment by TikhomirovSergey** [comment]
#594 #597

File: src/main/java/io/appium/java_client/remote/AppiumCommandExecutor.java [hunk 1]
```diff
@@ -47,124 +35,44 @@
...
-    public AppiumCommandExecutor(Map<String, AppiumCommandInfo> additionalCommands,
-        URL addressOfRemoteServer, HttpClient.Factory httpClientFactory) {
-        checkNotNull(addressOfRemoteServer);
-        remoteServer = addressOfRemoteServer;
-        this.additionalCommands = additionalCommands;
-        this.client = httpClientFactory.createClient(remoteServer);
+    public AppiumCommandExecutor(Map<String, CommandInfo> additionalCommands,
+                                 URL addressOfRemoteServer, HttpClient.Factory httpClientFactory) {
+        super(additionalCommands, addressOfRemoteServer, httpClientFactory);
+        service = null;
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/remote/AppiumCommandExecutor.java]: can this be a reason of unexpected NPE?

**Comment by mykola-mokhnach** [comment]
How about having some unit test to verify resulting capabilities structure?

File: src/main/java/io/appium/java_client/remote/AppiumCommandExecutor.java (hunk 1 above)
Comment (TikhomirovSergey) [inline src/main/java/io/appium/java_client/remote/AppiumCommandExecutor.java]: No. It is checked inside this method
https://github.com/TikhomirovSergey/java-client/blob/12cb9d9b6e117274340e8c73f40b0fa7a11396ba/src/main/java/io/appium/java_client/remote/AppiumCommandExecutor.java#L64

Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/remote/AppiumCommandExecutor.java]: then it might be useful to check the property as nullable

Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/remote/AppiumCommandExecutor.java]: or make it optional (I'd prefer this option)

**Comment by TikhomirovSergey** [comment]
@mykola-mokhnach 
>How about having some unit test to verify resulting capabilities structure?

I think we don't need for these tests:
1) as you can see `public class AppiumCommandExecutor extends HttpCommandExecutor` and

```java
 @Override public Response execute(Command command) throws IOException, WebDriverException {
        ...
      return super.execute(command);
       ...
   }
```

2) The capability forming is fully inherited from Selenium
[HttpCommandExecutor](https://github.com/SeleniumHQ/selenium/blob/master/java/client/src/org/openqa/selenium/remote/HttpCommandExecutor.java#L124) [ProtocolHandshake](https://github.com/SeleniumHQ/selenium/blob/master/java/client/src/org/openqa/selenium/remote/ProtocolHandshake.java)

Our tests which run some AppiumDriver are successful.

File: src/main/java/io/appium/java_client/remote/AppiumCommandExecutor.java (hunk 1 above)
Comment (TikhomirovSergey) [inline src/main/java/io/appium/java_client/remote/AppiumCommandExecutor.java]: @mykola-mokhnach ok. I will improve it soon.

File: src/main/java/io/appium/java_client/remote/AppiumCommandExecutor.java [hunk 2]
```diff
@@ -47,131 +37,56 @@
...
-
-            Response response = responseCodec.decode(httpResponse);
-            if (response.getSessionId() == null) {
-                if (httpResponse.getTargetHost() != null) {
-                    response.setSessionId(HttpSessionId.getSessionId(httpResponse.getTargetHost()));
-                } else {
-                    response.setSessionId(command.getSessionId().toString());
+    @Override public Response execute(Command command) throws WebDriverException {
+        if (DriverCommand.NEW_SESSION.equals(command.getName())) {
+            ofNullable(service).ifPresent(driverService -> {
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/remote/AppiumCommandExecutor.java]: I assume the code will be simpler if we make the service variable itself of type Optional

File: src/main/java/io/appium/java_client/remote/AppiumCommandExecutor.java [hunk 3]
```diff
@@ -47,131 +37,56 @@
...
-                    throw new WebDriverException("The appium server has accidentally died!", t);
-                }
+                    && rootCause.getMessage().contains("Connection refused")) {
+                throw ofNullable(service).map((Function<DriverService, WebDriverException>) service -> {
+                    if (service.isRunning()) {
+                        return new WebDriverException("The session is closed!", rootCause);
+                    }
+
+                    return new WebDriverException("The appium server has accidentally died!", rootCause);
+                }).orElse(new WebDriverException(rootCause.getMessage(), rootCause));
```
Comment (mykola-mokhnach) [inline src/main/java/io/appium/java_client/remote/AppiumCommandExecutor.java]: probably, orElseGet ?

**Comment by TikhomirovSergey** [comment]
@mykola-mokhnach There is another one improvement.

**Comment by mykola-mokhnach** [comment]
@TikhomirovSergey 👍

**Pull request merged**